In [1]:
# importing major libraries

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import cufflinks as cf
from plotly.offline import init_notebook_mode,download_plotlyjs
cf.go_offline(connected=True)

# importing additional libraries
import warnings
warnings.filterwarnings('ignore')

In [15]:
df = pd.read_csv('JMG_data.csv',nrows=None)

# about the dataset
This dataset contains records of vehicle listings from Craigslist contains most all relevant information that Craigslist provides on car sales such as price, condition, manufacturer, and other attributes.

In [16]:
df.columns

Index(['CarID', 'Listed_Price', 'Listed_Date', 'Make', 'Model', 'Year',
       'Vehicle_Type', 'Size', 'Color', 'Transmission', 'Fuel_Type', 'Drive',
       'Cylinders', 'Odometer', 'Condition', 'Title_Status', 'State',
       'Region'],
      dtype='object')

### Column Descriptions

- **Car ID**: An automatically generated index column.  
- **Listed_Price**: The listed price of the vehicle in USD.  
- **Listed_Date**: The date when the vehicle listing was posted.  
- **Make**: The name of the vehicle’s manufacturer.  
- **Model**: The specific model name of the vehicle.  
- **Year**: The build year of the vehicle.  
- **Vehicle_Type**: The classification of the vehicle.  
- **Size**: The size category of the vehicle.  
- **Color**: The exterior colour of the vehicle.  
- **Transmission**: The type of transmission (e.g., automatic, manual).  
- **Fuel_Type**: The type of energy the vehicle uses (e.g., gas, diesel, hybrid, electric).  
- **Drive**: The drivetrain configuration (e.g., FWD, RWD, AWD).  
- **Cylinders**: The number of engine cylinders.  
- **Odometer**: The vehicle’s mileage in miles.  
- **Condition**: The condition of the vehicle provided by the seller.  
- **Title_Status**: The legal status of the vehicle’s title.  
- **State**: The state abbreviation where the vehicle is listed.  
- **Region**: The region where the vehicle is listed for sale.  

In [17]:
# overview of data

df.head()

CarID  Listed_Price               Listed_Date       Make           Model  \
0  1246014         13987  2021-04-24T09:23:15-0500  chevrolet  silverado 1500   
1   692370          2800  2021-04-15T22:07:43-0400     toyota         4runner   
2   242958          1750  2021-05-02T13:02:12-0600       ford  escape xlt awd   
3   310455          9200  2021-04-12T12:20:38-0400      dodge         journey   
4   800040          9900  2021-04-30T13:36:13-0400     toyota        rav4 awd   

   Year Vehicle_Type       Size   Color Transmission Fuel_Type Drive  \
0  2007        truck   mid-size     red    automatic       gas   rwd   
1  1999          SUV  full-size  silver    automatic       gas   rwd   
2  2003          SUV   mid-size    grey    automatic       gas   4wd   
3  2015          SUV  full-size   white    automatic       gas   fwd   
4  2010          SUV   mid-size    grey    automatic       gas   4wd   

   Cylinders  Odometer  Condition Title_Status State                Region  
0        8.0    112709  excellent        clean    wi  appleton-oshkosh-FDL  
1        6.0    297053       good        clean    nc             asheville  
2        6.0    142500       good        clean    co                pueblo  
3        6.0    111000  excellent        clean    fl                 ocala  
4        4.0    112000       good        clean    ny               buffalo

In [21]:
df.shape

(62946, 18)

In [18]:
# converting Listed_Date into Date Time
df['Listed_Date'] = pd.to_datetime(df['Listed_Date'])

In [19]:
# seeking decription of the dataset
df.describe()

CarID  Listed_Price          Year     Cylinders      Odometer
count  6.294600e+04  62946.000000  62946.000000  62572.000000  6.294600e+04
mean   6.447435e+05  12972.070410   2008.470642      5.964505  1.250213e+05
std    3.607982e+05  11740.560353      9.892727      1.651513  2.424812e+05
min    9.300000e+01      0.000000   1900.000000      3.000000  0.000000e+00
25%    3.366578e+05   4995.000000   2006.000000      4.000000  7.404250e+04
50%    6.484725e+05   9000.000000   2011.000000      6.000000  1.149170e+05
75%    9.416182e+05  17500.000000   2014.000000      8.000000  1.560000e+05
max    1.280508e+06  75000.000000   2022.000000     12.000000  1.000000e+07


### **Interpretation & Insights**

- **Listed_Price**  
  The average listed price is around 13,000. Most vehicles are priced between 5,000 and 17,500. There are **entries with a price of 0 or as high as 75,000, indicating possible outliers** or data entry issues.

- **Year**  
  The vehicle manufacturing years range from 1900 to 2022. The **lower extreme (1900) is likely an error.** Most vehicles fall between 2006 and 2014, with the median year being 2011.

- **Cylinders**  
  Most cars have between 4 and 8 cylinders. Some listings have **unusual values like 3 or 12 cylinders.** A few values are missing, which may need to be handled during data preprocessing.

- **Odometer**  
  The mileage values **range from 0 to 10 million miles.** The median mileage is approximately 115,000 miles. **Extremely high values suggest the presence of outliers.** Entries with 0 mileage likely represent missing or incorrect data.

In [20]:
# seeking information from the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62946 entries, 0 to 62945
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   CarID         62946 non-null  int64  
 1   Listed_Price  62946 non-null  int64  
 2   Listed_Date   62946 non-null  object 
 3   Make          62946 non-null  object 
 4   Model         62946 non-null  object 
 5   Year          62946 non-null  int64  
 6   Vehicle_Type  62946 non-null  object 
 7   Size          62946 non-null  object 
 8   Color         62946 non-null  object 
 9   Transmission  62946 non-null  object 
 10  Fuel_Type     62946 non-null  object 
 11  Drive         62946 non-null  object 
 12  Cylinders     62572 non-null  float64
 13  Odometer      62946 non-null  int64  
 14  Condition     62946 non-null  object 
 15  Title_Status  62946 non-null  object 
 16  State         62946 non-null  object 
 17  Region        62645 non-null  object 
dtypes: float64(1), int64(4), o

### **Interpretation & Insights**

- **Dataset Size and Shape**  
  The dataset contains **62,946 records** and **18 columns**.

- **Data Types**  
  - Majority of the columns are of **object (categorical)** type (13 out of 18), indicating the need for **categorical encoding** before model training.
  - Numerical columns include: `Listed_Price`, `Year`, `Cylinders`, and `Odometer`.

- **Missing Values**  
  - `Cylinders` has **374 missing entries**, which could be imputed using median/mode or predicted based on other vehicle specs.
  - `Region` has **301 missing entries**, which might be recoverable from state or could be labeled as 'Unknown'.


In [12]:
# seeking duplicate entries

df.duplicated().sum()

np.int64(0)

In [13]:
# seeking null values in our dataset

df.isnull().sum()

CarID            0
Listed_Price     0
Listed_Date      0
Make             0
Model            0
Year             0
Vehicle_Type     0
Size             0
Color            0
Transmission     0
Fuel_Type        0
Drive            0
Cylinders       89
Odometer         0
Condition        1
Title_Status     1
State            1
Region          69
dtype: int64

**Missing Values Overview:**

- The dataset contains **89 missing entries** in the **`Cylinders`** column.  
- There is **1 missing value each** in the **`Condition`**, **`Title_Status`**, and **`State`** columns.  
- The **`Region`** column has **69 missing entries**.

These missing values are relatively few compared to the total dataset size (62,946 entries)

In [14]:
# dropping null values

df.dropna()

CarID  Listed_Price                Listed_Date       Make  \
0      1246014         13987  2021-04-24 09:23:15-05:00  chevrolet   
1       692370          2800  2021-04-15 22:07:43-04:00     toyota   
2       242958          1750  2021-05-02 13:02:12-06:00       ford   
3       310455          9200  2021-04-12 12:20:38-04:00      dodge   
4       800040          9900  2021-04-30 13:36:13-04:00     toyota   
...        ...           ...                        ...        ...   
15147   708201         34900  2021-04-19 10:26:14-04:00        ram   
15148   636045         16750  2021-04-12 16:19:42-05:00     subaru   
15149   543591          5000  2021-05-03 12:46:38-04:00  chevrolet   
15150   559209             0  2021-05-02 05:42:13-04:00      rover   
15151   501417         17000  2021-04-29 13:03:09-05:00       ford   

                      Model  Year Vehicle_Type       Size   Color  \
0            silverado 1500  2007        truck   mid-size     red   
1                   4runner  1999          SUV  full-size  silver   
2            escape xlt awd  2003          SUV   mid-size    grey   
3                   journey  2015          SUV  full-size   white   
4                  rav4 awd  2010          SUV   mid-size    grey   
...                     ...   ...          ...        ...     ...   
15147                  3500  2012        truck  full-size   white   
15148  outback 2.5i premium  2017        wagon  full-size     red   
15149        sonic lt turbo  2013        sedan    compact   black   
15150     evoque se premium  2017          SUV   mid-size  silver   
15151                 f-550  2004        truck  full-size   white   

      Transmission Fuel_Type Drive  Cylinders  Odometer  Condition  \
0        automatic       gas   rwd        8.0    112709  excellent   
1        automatic       gas   rwd        6.0    297053       good   
2        automatic       gas   4wd        6.0    142500       good   
3        automatic       gas   fwd        6.0    111000  excellent   
4        automatic       gas   4wd        4.0    112000       good   
...            ...       ...   ...        ...       ...        ...   
15147    automatic    diesel   4wd        6.0     85622  excellent   
15148    automatic       gas   4wd        4.0     35646  excellent   
15149    automatic       gas   fwd        4.0    151000  excellent   
15150    automatic       gas   4wd        4.0        43        new   
15151    automatic    diesel   rwd        8.0    130572       fair   

      Title_Status State                Region  
0            clean    wi  appleton-oshkosh-FDL  
1            clean    nc             asheville  
2            clean    co                pueblo  
3            clean    fl                 ocala  
4            clean    ny               buffalo  
...            ...   ...                   ...  
15147        clean    nc          fayetteville  
15148      rebuilt    mn     duluth / superior  
15149        clean    md             baltimore  
15150        clean    ma                boston  
15151        clean    ks              lawrence  

[14995 rows x 18 columns]